python/paddle/fluid/backward.py

```diff
@@ -1101,67 +1102,119 @@ def append_backward(loss,
         isinstance(callbacks, list)
 
     program = loss.block.program
+
+    # Todo(liym27): when to add 1 in control flow
     program._appending_grad_times += 1
 
     if no_grad_set is None:
         no_grad_set = set()
     no_grad_set = copy.copy(no_grad_set)
     no_grad_dict = _get_stop_gradients_(program)
+    # no_grad_set only contain vars in block 0
+    # Todo(liym27): support vars in sub block
     no_grad_dict[0].update(list(map(_append_grad_suffix_, no_grad_set)))
 
     root_block = program.block(0)
 
-    fwd_op_num = root_block.desc.op_size()
     current_block_idx = program.current_block_idx
+
+    current_block = program.block(current_block_idx)
+
+    # create grad block if in control flow
+    is_in_control_flow = current_block_idx != 0
+    if is_in_control_flow:
+        target_grad_block = program._create_block(
+            parent_idx=current_block.parent_idx)
+        target_grad_block._set_forward_block_idx(current_block_idx)
+        # after _create_block, program.current_block changes
+    else:
+        target_grad_block = root_block
+
+    def _get_son_parent_block_idx_dict(program, current_block_idx):
```

Comment:
Make this _get_son_parent_block_idx_dict function out of append_backward function